### DISCUSSION TITLE
SSG Data Revalidating NextJS v14

### DISCUSSION THREAD

**Discussion by VishalP-spinx** [vercel/next.js#66174] [Help]
### Summary

Hi, I would like to refresh the data on my statically generated page after a while.

We implemented a code for SSG according to the documentation. We use the generateStaticParams() for the SSG of the dynamic routes. Please let us know how we can revalidate that page data after some time.

**Note: We're using an APP router method.** 

We've implemented a fetch function to globalize and use the function to pass queries and get the data from the GQL response. You can check it out below that function code.

```
export const fetchData = async ({
  cache = "force-cache",
  headers,
  query,
  tags,
  variables,
  revalidate,
}: {
  cache?: RequestCache;
  headers?: HeadersInit;
  query?: string;
  tags?: string[];
  variables?: Record<string, any>;
  revalidate?: number;
}): Promise<{ status: number; data: any }> => {
  try {
    let config = {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        ...headers,
      },
      body: JSON.stringify({
        ...(query && { query }),
        ...(variables && { variables }),
      }),
      ...(!revalidate && {cache} ),
      ...(tags && { next: { tags } }),
      ...(revalidate && { next: { revalidate: revalidate } }),
    };

    const result = await fetch(process.env.GQL_URL!, config);

    const { data } = await result.json();

    if (data.errors) {
      throw data?.errors[0]?.message;
    }

    return {
      status: result.status,
      data,
    };
  } catch (error) {
    if (typeof error === "string") {
      throw new Error(error);
    } else {
      throw new Error("An unknown error occurred");
    }
  }
};

```
**Please help me out.**

**Comment by PetrovaDaria** [comment]
Maybe you can use `export const revalidate = /* your number of seconds after which you want to update pages */`. It will 
revalidate full route cache for these ssg pages. https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#revalidate
I'm not an expert and right now I'm at the point where I want to try to use this parameter in my project but it seems that we have the same situation. Except the fact that you use `fetch` which may be revalidated several other ways.

**Reply by TimGeudens** [reply]
We have the same scenario:

- Statically generated page at build time
- fetch calls with cache tags
- revalidation endpoint to revalidate those fetch calls

We would like to revalidate 1 specific fetch call every 15 minutes. So I tried adding the revalidate parameter to the fetch call:
`next: { revalidate: 20 }`
However that does not seem to be working. 

Next thing I tried is adding the revalidate param to the full route so our code looks like this:

`export const generateStaticParams = async () => [];
export const revalidate = 10;`

But the statically generated page is never regenerated after the initial build. It looks to me like the revalidate and generateMetaData params are not working in combination. 

Our only other solution it this time seems like revalidating the fetch cache tag with a cron job or something like that, but seems a bit overkill. The NextJS docs are not clear about using those 2 parameters in combo, maybe someone can chime in and confirm that is the case?

**Reply by fenggwsx** [reply]
I suppose you‘re getting the time unit wrong here. The unit of `revalidate` property here is second. If you want to revalidate every 15  minutes, the `revalidate` value should be 900.

The fetch call should look like this:

```typescript
const res = await fetch(
    `https://xxx.example.com/${params.id}`,
    { headers: { Accept: "application/json" }, next: { revalidate: 900 } }
);
```

I'm now using Next.js 14.2.5, and this works for me in a dynamic route. By the way, I used a time-based API to test the revalidation.

**Reply by TimGeudens** [reply]
Works with dynamic route indeed, but not an SSG route. the revalidate value of 20 seconds was just a value to test the functionality.
